clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -283,6 +283,178 @@ void Sema::checkSYCLDeviceVarDecl(VarDecl *Var) {
+                 const Stmt *St) {
+  // We have a variable declaration and want to know if it is referenced
+  // in this statment or its children, if any.
+  if (!St)
+    return false;
+
+  const Expr *E = dyn_cast<Expr>(St);
+  if (E)
+    return isVarInExpr(V, AssignF, E);
+  else {
+    for (const Stmt *SubStmt : St->children()) {
+      bool match = isVarInStmt(V, AssignF, SubStmt);
+      if (match)
+        return match;
+    }
+  }
+  return false;
+}
+
+void Sema::checkSYCLDevicePointerCapture(VarDecl *Var,
+                                         SourceLocation CaptureLoc) {
+  // Any pointer captured into the SYCL kernel lambda will fail when
+  // dereferenced...except USM. If it weren't for USM we could just emit a
+  // deferred diagnostic for every pointer capture. Instead, we attempt to
+  // identify which pointers are USM, and which are definitely not (and will
+  // crash). For those that will crash, we emit an error. For those that are
+  // unknown, we _could_ emit a gentle note suggesting the user double check
+  // that they are using USM.  But at this time we do not. For safe USM
+  // pointers, we do nothing.
+
+  enum ExprAllocation { Unknown, USM, WillCrash };
+  ExprAllocation howAllocated = Unknown;
+
+  // We will use this to check declarations and assignments.
+  auto VetteCallExpr = [&howAllocated](const Expr *E) {
+    E = E->IgnoreCasts();
+    const CallExpr *CE = dyn_cast<CallExpr>(E);
+    bool updated = false;
+    if (CE) {
+      const FunctionDecl *func = CE->getDirectCallee();
+      auto FullName = func->getQualifiedNameAsString();
+      // Check to see if this function call is one of the USM allocators.
+      if ((FullName.rfind("cl::sycl::malloc", 0) == 0) ||
+          (FullName.rfind("cl::sycl::aligned_alloc", 0) == 0)) {
+        howAllocated = USM;
+        updated = true;
+      } else if (howAllocated == Unknown &&
+                 ((FullName.compare("malloc") == 0) ||
+                  (FullName.compare("calloc") == 0))) {
+        howAllocated = WillCrash;
+        updated = true;
+      }
+    }
+    return updated;
+  };
+  std::function<bool(const Expr *)> AssignF = VetteCallExpr;
+
+  SourceLocation DecLoc = SourceLocation();
+
+  // Try to determine provenance of this Var.
+  const Expr *Init = Var->getAnyInitializer();
+  if (Init) {
+    Init = Init->IgnoreCasts();
+    DecLoc = Init->getExprLoc();
+
+    const CallExpr *CE = dyn_cast<CallExpr>(Init);
+    if (CE) {
+      VetteCallExpr(CE); // Modifies howAllocated via side-effect.
+    } else {
+      // Var has initialization, but not as return result of a function,
+      // disqualify any other obvious bad initialization expressions.
+      const StringLiteral *SL = dyn_cast<StringLiteral>(Init);
+      if (SL)
+        howAllocated = WillCrash;
+
+      const UnaryOperator *UO = dyn_cast<UnaryOperator>(Init);
+      if (UO && (UO->getOpcode() == UO_AddrOf))
```

Comment:
I'm guessing this logic is going to be quite a bit of false errors.  You cannot really do that.

Reply:
This is only screening variable initialization (of pointers we want to capture in the kernel lambda), it's not forbidding the use of the address_of operator generally.
 If a pointer variable was initialized by taking the address of another variable, that's not going to be capturable. That will segfault.     So, once the other analysis is done and we are confident the variable wasn't otherwise referenced, we can emit the error in confidence.  Am I overlooking an obvious counter-example?

Reply:
you are correct. Removed.